api/src/wfl/log.clj

```diff
@@ -1,45 +1,97 @@
+
+(s/def ::level-string   (s/and  string? level-string?))
+(s/def ::level-request  (s/keys :req-un [::level-string]))
+(s/def ::level-response (s/keys :req-un [::level-string]))
+
+(def ^:private active-map
+  "Map a level keyword to a set of active levels."
+  (loop [sofar {} levels levels]
+    (if-let [level (first levels)]
+      (recur (assoc sofar level (set levels)) (rest levels))
+      sofar)))
+
+(def ^:private active-level-predicate
+  "The active level predicate now."
+  (atom (:error active-map)))
+
+(defn set-active-level
+  "Set `active-level-predicate` for the `level` string."
+  [level]
+  (reset! active-level-predicate
+          (-> (if (empty? level) "info" level)
+              str/lower-case keyword active-map)))
+
+;; https://cloud.google.com/logging/docs/agent/logging/configuration#special-fields
+;;
+(def ^:private google-fields
+  "Map WFL's log field names to what Stackdriver recognizes."
+  {::httpRequest    :httpRequest
+   ::insertId       :logging.googleapis.com/insertId
+   ::labels         :logging.googleapis.com/labels
+   ::message        :message
+   ::operation      :logging.googleapis.com/operation
+   ::severity       :severity
+   ::sourceLocation :logging.googleapis.com/sourceLocation
+   ::spanId         :logging.googleapis.com/spanId
+   ::time           :time
+   ::trace          :logging.googleapis.com/trace})
 
 (defn ^:private key-fn
   "Preserve the namespace of `key` when qualified."
   [key]
   ((:key-fn json/default-write-options)
-   (if (qualified-keyword? key)
-     (str (namespace key) \/ (name key))
-     key)))
+   (let [googled (google-fields key key)]
+     (if (qualified-keyword? googled)
+       (str (namespace googled) \/ (name googled))
+       key))))
 
-;; HACK: Override how JSONWriter handles Throwables.
-;; json/write-object is private and handles java.util.Map.
-;;
+;; HACK: Override how JSONWriter handles some EDN values.
+;; Not sure what to do about java.lang.Class and such.
+
+(defn ^:private write-tagged
+  "Write the TaggedLiteral X to OUT as JSON with OPTIONS."
+  [x out options]
+  (#'json/write-string (pr-str x) out options))
+(defn ^:private write-character
+  "Write the Character X to OUT as JSON with OPTIONS."
+  [x out options]
+  (#'json/write-string (str x) out options))
+(defn ^:private write-class
+  "Write the Class X to OUT as JSON with OPTIONS."
+  [_x out options]
+  (#'json/write-string "(-Some java.lang.Class!-)" out options))
 (defn ^:private write-throwable
   "Write the Throwable X to OUT as JSON with OPTIONS."
   [x out options]
-  (#'json/write-object (Throwable->map x) out (assoc options
-                                                     :escape-slash false
-                                                     :key-fn       key-fn)))
-
-(extend java.lang.Throwable json/JSONWriter {:-write write-throwable})
+  (#'json/write-object (Throwable->map x) out options))
+(extend clojure.lang.ExceptionInfo json/JSONWriter {:-write write-tagged})
```

Comment:
Thanks.

I "unfinished" the local staged workload with a deleted workspace that I examined in https://github.com/broadinstitute/wfl/pull/568

The root cause error is now appropriately logged and registered as an update loop failure for that specific workload, but does not crash the whole update loop.

[exceptioninfo-workload-failure.txt](https://github.com/broadinstitute/wfl/files/8063251/exceptioninfo-workload-failure.txt)